src/zope/file/download.py

```diff
@@ -56,8 +56,12 @@ def getHeaders(context, contentType=None, downloadName=None,
         downloadName = downloadName or context.__name__
         if contentDisposition:
             if downloadName:
+                # Headers must be native strings. Under Py2, we probably
+                # need to encode the name
+                if str is bytes and not isinstance(downloadName, bytes):
+                    downloadName = downloadName.encode("utf-8")
                 contentDisposition += (
-                    '; filename="%s"' % downloadName.encode("utf-8")
+                    '; filename="%s"' % downloadName
```

Comment:
IIRC [PEP-3333](https://www.python.org/dev/peps/pep-3333/#unicode-issues) (WSGI 1.0.1) officially enshrines doubly-encoded strings(e.g. UTF-8 encoded to Latin-1), essentially using Latin-1-only unicode strings instead of bytestrings on Python 3.

What I wasn't sure about was whether we're dealing with WSGI-level strings here, or if any of the numerous HTTP request/response abstraction levels we have here deal with Unicode conversions for us.